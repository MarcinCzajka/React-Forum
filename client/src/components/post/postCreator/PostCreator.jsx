import React from 'react';
import { Helmet } from "react-helmet";
import { Button, Statistic, Icon } from "semantic-ui-react";
import UserContext from '../../../contexts/UserContext';
import { LocaleConsumer } from '../../../contexts/LocaleContext';
import ImageModal from '../../imageModal/ImageModal';
import ErrorMessage from '../../message/ErrorMessage';
import { createForumPost } from '../forumPost/forumPostLogic/forumPostApi';
import '../forumPost/layout/ForumPost.scss';
import './PostCreator.scss';

class PostCreator extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
			title: '',
			shortDescription: '',
			description: '',
			category: 'General',
			image: '',
            error: ''
        }

        this.imageModal = React.createRef();
    }

    static contextType = UserContext;

    componentDidMount() {
        this.downloadCloudinaryWidget();
    }

    showImageModal = () => {
        if(this.state.image) this.imageModal.current.open();
    }

    createNewForumPost = () => {
        const { title, description, category, image } = this.state;
        createForumPost(this.context.userId, title, description, category, image)
            .then((res) => {
                if (res.status === 200) {
                    //Change location to new post location
                    window.location = window.location.origin + '/post/' + res.data._id;
                }
            })
            .catch(err => {
                this.setState({ error: err.response.data });
            })
    }

    focus = (className) => {
        if(typeof className !== 'string') return
        
        document.getElementsByClassName(className)[0].focus();
    }

    createWidget = () => {
        this.cloudinaryWidget = window.cloudinary.createUploadWidget({
            cloudName: 'dswujhql5', uploadPreset: 'ot93kwr6',
            multiple: false,
            sources: [ 'local', 'url', 'camera' ],
            maxFileSize: 5000000 }, (error, result) => { 
              if (!error && result && result.event === "success") { 

                this.setState({image: result.info.secure_url});
                this.cloudinaryWidget.close();
              }
            }
        )
    }

    showWidget = () => {
        this.cloudinaryWidget.open();
    }

    downloadCloudinaryWidget = () => {
        const script = document.createElement('script');
        script.addEventListener('load', this.createWidget);
        script.async = true;
        script.defer = true;
        script.src = "https://widget.cloudinary.com/v2.0/global/all.js";

        document.getElementsByTagName('body')[0].appendChild(script);
    }

    changeTitle = (e) => {
        if(e.target.value.length >= 50) return;
        
        this.setState({title: e.target.value})
    }

    render() {
        const { title, description, image, error } = this.state;
        return (
            <LocaleConsumer>
                {locale => (
                    <>
                        <Helmet>
                            <title>{locale.postCreator.title} - {locale.appName}</title>
                        </Helmet>

                        <article className='roomContainer postContainerOverlay'>

                            <div>
                                {!this.context.loggedIn ? (
                                    <ErrorMessage 
                                        message={<p>Only <span className='asLink' onClick={this.context.showLogin}>logged in</span> users can post new stuff.</p>}  
                                    />
                                ) : error ? (
                                    <ErrorMessage message={error} />
                                ) : ''}
                            </div>

                            <div className='roomGrid noMargin noPadding' >

                                    <div className='roomImageContainer imageContainerInCreation' >
                                        <img 
                                            className='roomImage'
                                            src={image}
                                            alt={title}
                                        />
                                        <div className='imageOverlay imageOverlayInCreation' >
                                            <Icon id='zoomIcon' name='zoom-in' inverted size='huge' onClick={this.showImageModal} />
                                            <Icon id='uploadIcon' name='file outline image' size='huge' inverted onClick={this.showWidget}></Icon>
                                        </div>
                                    </div>

                                <header className='roomTitle roomTitleCreator noMargin noPadding' onClick={this.editTitle}>
                                    <textarea 
                                        className='titleInput'
                                        value={title} 
                                        placeholder={locale.postCreator.changeTitlePlaceholder}
                                        onChange={this.changeTitle}>
                                    </textarea>

                                    <div className='titleOverlay' onClick={() => this.focus('titleInput')}>
                                        <Icon name='edit outline'></Icon>
                                    </div>
                                </header>

                                <main className='roomDescription noMargin noPadding' onClick={this.editDescription}>
                                    <textarea 
                                        className='descriptionInput'
                                        value={description} 
                                        placeholder={locale.postCreator.changeContentPlaceholder}
                                        onChange={e => {this.setState({description: e.target.value})}}>
                                    </textarea>

                                    <p style={{color: (description.length <= 25)}}>
                                        {500 - description.length} letters left.
                                    </p>

                                    <div className='descriptionOverlay' onClick={() => this.focus('descriptionInput')}>
                                        <Icon name='edit outline' size='big'></Icon>
                                    </div>
                                </main>

                                <footer className='roomFooter'>
                                    <Statistic.Group size='mini' className='maxWidth roomStats noMargin'>
                                        <Statistic className='roomStat'>
                                            <Statistic.Value>
                                                <Icon name='comments outline'> 0</Icon>
                                            </Statistic.Value>
                                        </Statistic>

                                        <Statistic className='roomStat'>
                                            <Statistic.Value><Icon name='eye'/>  0</Statistic.Value>
                                        </Statistic>
                                        

                                        <Statistic className='roomStat' >
                                            <Statistic.Value>
                                                <Icon name='thumbs up' />  0
                                            </Statistic.Value>
                                        </Statistic>
                                    </Statistic.Group>
                                </footer>

                                <ImageModal
                                    image={image}
                                    alt={title}
                                    ref={this.imageModal}
                                />

                            </div>

                        </article>
                                
                        <Button 
                            className='createPostBtn' 
                            icon labelPosition='right' 
                            color='blue' 
                            onClick={this.createNewForumPost} 
                        >
                            {locale.postCreator.createPostButton}
                            <Icon name='paper plane' />
                        </Button>
                    </>
                )}
            </LocaleConsumer>
        )
    }

}

export default PostCreator;