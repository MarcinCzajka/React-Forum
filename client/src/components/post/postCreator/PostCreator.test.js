import React from 'react';
import { shallow } from 'enzyme';
import { create } from "react-test-renderer";
import PostCreator from './PostCreator';
import { UserProvider } from '../../../contexts/UserContext';
import { LocaleProvider } from '../../../contexts/LocaleContext';

const locale = {
    appName: 'abedef',
    postCreator: {
        title: 'abcde',
        changeTitlePlaceholder: 'abcde',
        changeContentPlaceholder: 'fdsfds',
        createPostButton: 'afdfsddf'
    }
}

describe('PostListPagination component', () => {
    test('renders without crashing', () => {
        shallow(<LocaleProvider value={locale}><PostCreator /></LocaleProvider>);
    });

    test('match snapshot w/o locale and userContext', () => {
        const component = create(<LocaleProvider value={locale}><PostCreator /></LocaleProvider>);
        expect(component).toMatchSnapshot();
    });

    test('match snapshot /w locale and w/o userContext', () => {
        const component = create(<LocaleProvider value={locale}><PostCreator /></LocaleProvider>);
        expect(component).toMatchSnapshot();
    });

    test('match snapshot /w locale and user logged in', () => {
        const component = create(<LocaleProvider value={locale}>
            <UserProvider value={{loggedIn: true}} >
                <PostCreator />
            </UserProvider>
        </LocaleProvider>
    );
        expect(component).toMatchSnapshot();
    });

    test('match snapshot /w locale and user logged out', () => {
        const component = create(<LocaleProvider value={locale}>
            <UserProvider value={{loggedIn: false}} >
                <PostCreator />
            </UserProvider>
        </LocaleProvider>
    );
        expect(component).toMatchSnapshot();
    });

});