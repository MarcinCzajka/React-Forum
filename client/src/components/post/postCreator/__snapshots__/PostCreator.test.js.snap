// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostListPagination component match snapshot /w locale and user logged in 1`] = `
<article
  className="roomContainer"
>
  <div>
    
  </div>
  <div
    className="roomGrid noMargin noPadding"
  >
    <div
      className="roomImageContainer imageContainerInCreation"
    >
      <img
        alt=""
        className="roomImage"
        src=""
      />
      <div
        className="imageOverlay imageOverlayInCreation"
      >
        <i
          aria-hidden="true"
          className="zoom-in huge inverted icon"
          id="zoomIcon"
          onClick={[Function]}
        />
        <i
          aria-hidden="true"
          className="file outline image huge inverted icon"
          id="uploadIcon"
          onClick={[Function]}
        />
      </div>
    </div>
    <header
      className="roomTitle roomTitleCreator noMargin noPadding"
    >
      <textarea
        className="titleInput"
        onChange={[Function]}
        placeholder="abcde"
        value=""
      />
      <div
        className="titleOverlay"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="edit outline icon"
          onClick={[Function]}
        />
      </div>
    </header>
    <main
      className="roomDescription noMargin noPadding"
    >
      <textarea
        className="descriptionInput"
        onChange={[Function]}
        placeholder="fdsfds"
        value=""
      />
      <p
        style={
          Object {
            "color": true,
          }
        }
      >
        0
         letters left.
      </p>
      <div
        className="descriptionOverlay"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="edit outline big icon"
          onClick={[Function]}
        />
      </div>
    </main>
    <footer
      className="roomFooter"
    >
      <div
        className="ui mini statistics maxWidth roomStats noMargin"
      >
        <div
          className="ui statistic roomStat"
        >
          <div
            className="value"
          >
            <i
              aria-hidden="true"
              className="comments outline icon"
              onClick={[Function]}
            >
               0
            </i>
          </div>
        </div>
        <div
          className="ui statistic roomStat"
        >
          <div
            className="value"
          >
            <i
              aria-hidden="true"
              className="eye icon"
              onClick={[Function]}
            />
              0
          </div>
        </div>
        <div
          className="ui statistic roomStat"
        >
          <div
            className="value"
          >
            <i
              aria-hidden="true"
              className="thumbs up icon"
              onClick={[Function]}
            />
              0
          </div>
        </div>
      </div>
    </footer>
    <button
      className="ui blue icon right labeled button createPostBtn"
      onClick={[Function]}
    >
      afdfsddf
      <i
        aria-hidden="true"
        className="paper plane icon"
        onClick={[Function]}
      />
    </button>
  </div>
</article>
`;

exports[`PostListPagination component match snapshot /w locale and user logged out 1`] = `
<article
  className="roomContainer"
>
  <div>
    <div
      className="ui error message"
    >
      <div
        className="content"
      >
        <div
          className="header"
        >
          Error
        </div>
        <p>
          Only 
          <span
            className="asLink"
          >
            logged in
          </span>
           users can post new stuff.
        </p>
      </div>
    </div>
  </div>
  <div
    className="roomGrid noMargin noPadding"
  >
    <div
      className="roomImageContainer imageContainerInCreation"
    >
      <img
        alt=""
        className="roomImage"
        src=""
      />
      <div
        className="imageOverlay imageOverlayInCreation"
      >
        <i
          aria-hidden="true"
          className="zoom-in huge inverted icon"
          id="zoomIcon"
          onClick={[Function]}
        />
        <i
          aria-hidden="true"
          className="file outline image huge inverted icon"
          id="uploadIcon"
          onClick={[Function]}
        />
      </div>
    </div>
    <header
      className="roomTitle roomTitleCreator noMargin noPadding"
    >
      <textarea
        className="titleInput"
        onChange={[Function]}
        placeholder="abcde"
        value=""
      />
      <div
        className="titleOverlay"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="edit outline icon"
          onClick={[Function]}
        />
      </div>
    </header>
    <main
      className="roomDescription noMargin noPadding"
    >
      <textarea
        className="descriptionInput"
        onChange={[Function]}
        placeholder="fdsfds"
        value=""
      />
      <p
        style={
          Object {
            "color": true,
          }
        }
      >
        0
         letters left.
      </p>
      <div
        className="descriptionOverlay"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="edit outline big icon"
          onClick={[Function]}
        />
      </div>
    </main>
    <footer
      className="roomFooter"
    >
      <div
        className="ui mini statistics maxWidth roomStats noMargin"
      >
        <div
          className="ui statistic roomStat"
        >
          <div
            className="value"
          >
            <i
              aria-hidden="true"
              className="comments outline icon"
              onClick={[Function]}
            >
               0
            </i>
          </div>
        </div>
        <div
          className="ui statistic roomStat"
        >
          <div
            className="value"
          >
            <i
              aria-hidden="true"
              className="eye icon"
              onClick={[Function]}
            />
              0
          </div>
        </div>
        <div
          className="ui statistic roomStat"
        >
          <div
            className="value"
          >
            <i
              aria-hidden="true"
              className="thumbs up icon"
              onClick={[Function]}
            />
              0
          </div>
        </div>
      </div>
    </footer>
    <button
      className="ui blue icon right labeled button createPostBtn"
      onClick={[Function]}
    >
      afdfsddf
      <i
        aria-hidden="true"
        className="paper plane icon"
        onClick={[Function]}
      />
    </button>
  </div>
</article>
`;

exports[`PostListPagination component match snapshot /w locale and w/o userContext 1`] = `
<article
  className="roomContainer"
>
  <div>
    <div
      className="ui error message"
    >
      <div
        className="content"
      >
        <div
          className="header"
        >
          Error
        </div>
        <p>
          Only 
          <span
            className="asLink"
          >
            logged in
          </span>
           users can post new stuff.
        </p>
      </div>
    </div>
  </div>
  <div
    className="roomGrid noMargin noPadding"
  >
    <div
      className="roomImageContainer imageContainerInCreation"
    >
      <img
        alt=""
        className="roomImage"
        src=""
      />
      <div
        className="imageOverlay imageOverlayInCreation"
      >
        <i
          aria-hidden="true"
          className="zoom-in huge inverted icon"
          id="zoomIcon"
          onClick={[Function]}
        />
        <i
          aria-hidden="true"
          className="file outline image huge inverted icon"
          id="uploadIcon"
          onClick={[Function]}
        />
      </div>
    </div>
    <header
      className="roomTitle roomTitleCreator noMargin noPadding"
    >
      <textarea
        className="titleInput"
        onChange={[Function]}
        placeholder="abcde"
        value=""
      />
      <div
        className="titleOverlay"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="edit outline icon"
          onClick={[Function]}
        />
      </div>
    </header>
    <main
      className="roomDescription noMargin noPadding"
    >
      <textarea
        className="descriptionInput"
        onChange={[Function]}
        placeholder="fdsfds"
        value=""
      />
      <p
        style={
          Object {
            "color": true,
          }
        }
      >
        0
         letters left.
      </p>
      <div
        className="descriptionOverlay"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="edit outline big icon"
          onClick={[Function]}
        />
      </div>
    </main>
    <footer
      className="roomFooter"
    >
      <div
        className="ui mini statistics maxWidth roomStats noMargin"
      >
        <div
          className="ui statistic roomStat"
        >
          <div
            className="value"
          >
            <i
              aria-hidden="true"
              className="comments outline icon"
              onClick={[Function]}
            >
               0
            </i>
          </div>
        </div>
        <div
          className="ui statistic roomStat"
        >
          <div
            className="value"
          >
            <i
              aria-hidden="true"
              className="eye icon"
              onClick={[Function]}
            />
              0
          </div>
        </div>
        <div
          className="ui statistic roomStat"
        >
          <div
            className="value"
          >
            <i
              aria-hidden="true"
              className="thumbs up icon"
              onClick={[Function]}
            />
              0
          </div>
        </div>
      </div>
    </footer>
    <button
      className="ui blue icon right labeled button createPostBtn"
      onClick={[Function]}
    >
      afdfsddf
      <i
        aria-hidden="true"
        className="paper plane icon"
        onClick={[Function]}
      />
    </button>
  </div>
</article>
`;

exports[`PostListPagination component match snapshot w/o locale and userContext 1`] = `
<article
  className="roomContainer"
>
  <div>
    <div
      className="ui error message"
    >
      <div
        className="content"
      >
        <div
          className="header"
        >
          Error
        </div>
        <p>
          Only 
          <span
            className="asLink"
          >
            logged in
          </span>
           users can post new stuff.
        </p>
      </div>
    </div>
  </div>
  <div
    className="roomGrid noMargin noPadding"
  >
    <div
      className="roomImageContainer imageContainerInCreation"
    >
      <img
        alt=""
        className="roomImage"
        src=""
      />
      <div
        className="imageOverlay imageOverlayInCreation"
      >
        <i
          aria-hidden="true"
          className="zoom-in huge inverted icon"
          id="zoomIcon"
          onClick={[Function]}
        />
        <i
          aria-hidden="true"
          className="file outline image huge inverted icon"
          id="uploadIcon"
          onClick={[Function]}
        />
      </div>
    </div>
    <header
      className="roomTitle roomTitleCreator noMargin noPadding"
    >
      <textarea
        className="titleInput"
        onChange={[Function]}
        placeholder="abcde"
        value=""
      />
      <div
        className="titleOverlay"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="edit outline icon"
          onClick={[Function]}
        />
      </div>
    </header>
    <main
      className="roomDescription noMargin noPadding"
    >
      <textarea
        className="descriptionInput"
        onChange={[Function]}
        placeholder="fdsfds"
        value=""
      />
      <p
        style={
          Object {
            "color": true,
          }
        }
      >
        0
         letters left.
      </p>
      <div
        className="descriptionOverlay"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="edit outline big icon"
          onClick={[Function]}
        />
      </div>
    </main>
    <footer
      className="roomFooter"
    >
      <div
        className="ui mini statistics maxWidth roomStats noMargin"
      >
        <div
          className="ui statistic roomStat"
        >
          <div
            className="value"
          >
            <i
              aria-hidden="true"
              className="comments outline icon"
              onClick={[Function]}
            >
               0
            </i>
          </div>
        </div>
        <div
          className="ui statistic roomStat"
        >
          <div
            className="value"
          >
            <i
              aria-hidden="true"
              className="eye icon"
              onClick={[Function]}
            />
              0
          </div>
        </div>
        <div
          className="ui statistic roomStat"
        >
          <div
            className="value"
          >
            <i
              aria-hidden="true"
              className="thumbs up icon"
              onClick={[Function]}
            />
              0
          </div>
        </div>
      </div>
    </footer>
    <button
      className="ui blue icon right labeled button createPostBtn"
      onClick={[Function]}
    >
      afdfsddf
      <i
        aria-hidden="true"
        className="paper plane icon"
        onClick={[Function]}
      />
    </button>
  </div>
</article>
`;
